addon/lib/Feature.jsm

```diff
@@ -1,140 +1,616 @@
+        type: "request-aborted",
+        partial: measurement
+      });
+      req.onerror = () => sendMessageToParent("error", {
+        type: "request-error",
+        partial: measurement
+      });
+
+      req.send();
+    };
+
+    return "data:text/html,<meta charset='utf8'/><script>" +
+           encodeURIComponent(inferConnectionLabel.toSource()) +
+           "</script><script>(" +
+           encodeURIComponent(contentScript.toSource()) +
+           ")(" + endpointUrl.toSource() + ", " + DOWNLINK_SAMPLING_MS + ");</script>";
   }
 
-  /* no-op shutdown */
-  shutdown() {}
-}
+  /**
+   * Try to check if we're offline.
+   * @return {Boolean} true if we're offline or behind a captive portal, false
+   *         otherwise.
+   */
+  _offline() {
+    // Services.io.offline has slowly become fairly useless over the years - it
+    // no longer attempts to track the actual network state by default, but one
+    // thing stays true: if it says we're offline then we are definitely not online.
+    //
+    // We also ask the captive portal service if we are behind a locked captive
+    // portal.
+    //
+    // We don't check on the NetworkLinkService however, because it gave us
+    // false positives in the past in a vm environment.
+    try {
+      if (Services.io.offline ||
+          CaptivePortalService.state == CaptivePortalService.LOCKED_PORTAL) {
+        return true;
+      }
+    } catch (ex) {
+      this._log.warn("Could not determine network status.", ex);
+    }
+    return false;
+  }
 
+  /**
+   * This function is called during idle time windows to perform a measurement.
+   */
+  async _performMeasurement() {
+    this._log.debug("_performMeasurement");
 
+    // Did we expire while the study was running?
+    if (this.HasExpired()) {
+      await this._studyUtils.endStudy({ reason: "expired" });
+      return;
+    }
+
+    // If we're in the control group, do nothing.
+    if (this._variation.name === "control") {
+      this._log.debug("_performMeasurement - control group, bailing out");
+      return;
+    }
+
+    // Are we online? Do we have network connectivity?
+    if (this._offline()) {
+      this._log.debug("_performMeasurement - offline");
+      return;
+    }
+
+    // Check if enough time passed since the last download.
+    const lastMeasurementDate = getDateFromPref(this._studyPrefs.LastMeasurement);
+    if (lastMeasurementDate &&
+        Math.abs(lastMeasurementDate - Date.now()) < this._delayBetweenMeasurements) {
+      this._log.debug(`_performMeasurement - skipping idle (last measurement ${lastMeasurementDate})`);
+      return;
+    }
+
+    // Bail out if a measurement is already in progress.
+    if (this._isMeasuring) {
+      this._log.warn("_performMeasurement - measurement in progress");
+      return;
+    }
+
+    // Trigger the request.
+    let endpointUrl =
+      Services.prefs.getCharPref(this._studyPrefs.Endpoint.name, `${this._studyPrefs.Endpoint.defaultValue}`);
```

Comment:
No, the default value will be changed to the URI provided in [bug 1433927](https://bugzilla.mozilla.org/show_bug.cgi?id=1433927) once available. It will definitely be a string.